File: vikoba_app/test/loan_engine_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:vikoba_app/core/domain/vizoba_calc.dart';
import 'package:vikoba_app/models/loan.dart';
import 'package:vikoba_app/models/loan_product.dart';
import 'package:vikoba_app/models/loan_schedule.dart';
import 'package:vikoba_app/models/member.dart';

/// Mirrors `backend/tests/Unit/LoanEngineTest.php` — the Flutter copy of the
/// money rules must agree with the authoritative Laravel engine.
void main() {
  LoanProduct product({
    String name = 'Test Loan',
    double rate = 10,
    LoanInterestMethod method = LoanInterestMethod.flat,
    int maxTerm = 6,
    int multiplier = 4,
    double minAmount = 20000,
    PenaltyType penaltyType = PenaltyType.flat,
    double penaltyValue = 5000,
    int penaltyGrace = 0,
    int penaltyPeriod = 7,
  }) =>
      LoanProduct(
        id: 'P1',
        name: name,
        interestRate: rate,
        interestMethod: method,
        maxTermMonths: maxTerm,
        maxMultiplier: multiplier,
        minAmount: minAmount,
        penaltyType: penaltyType,
        penaltyValue: penaltyValue,
        penaltyGraceDays: penaltyGrace,
        penaltyPeriodDays: penaltyPeriod,
      );

  Member member({int shares = 40, int daysJoined = 120, bool active = true}) =>
      Member(
        id: 'MEM1',
        fullName: 'Amina Juma',
        phoneNumber: '0712345678',
        role: MemberRole.member,
        joinedDate: DateTime.now().subtract(Duration(days: daysJoined)),
        totalShares: shares,
        shareValue: 10000,
        isActive: active,
      );

  Loan loan({
    double principal = 100000,
    LoanProduct? prod,
    int term = 1,
    LoanStatus status = LoanStatus.active,
    Duration lateBy = Duration.zero,
    double penaltyAccrued = 0,
  }) {
    final p = prod ?? product();
    final schedules = VikobaCalc.amortize(
      principal: principal,
      ratePercent: p.interestRate,
      termMonths: term,
      method: p.interestMethod,
    );
    return Loan(
      id: 'L1',
      memberId: 'MEM1',
      memberName: 'Amina Juma',
      principal: principal,
      interestRate: p.interestRate,
      loanProductId: p.id,
      termMonths: term,
      interestMethod: p.interestMethod,
      issuedDate: DateTime.now(),
      dueDate: DateTime.now().subtract(lateBy),
      status: status,
      penaltyAccrued: penaltyAccrued,
      schedules: schedules,
    );
  }

  double sumPrincipal(List<LoanSchedule> rows) =>
      rows.fold(0.0, (s, r) => s + r.principalDue);

  group('amortize (mirrors backend)', () {
    test('flat 100k @10% for 3 months sums exactly with drift to earliest', () {
      final rows = VikobaCalc.amortize(
          principal: 100000, ratePercent: 10, termMonths: 3);

      expect(rows.length, 3);
      expect(sumPrincipal(rows), 100000);
      expect(rows.fold(0.0, (s, r) => s + r.interestDue), 30000);
      expect(rows.first.principalDue, 33334);
      expect(rows[1].principalDue, 33333);
      expect(rows[2].principalDue, 33333);
    });

    test('single flat installment matches the legacy single-cycle formula', () {
      final rows = VikobaCalc.amortize(
          principal: 100000, ratePercent: 10, termMonths: 1);
      expect(rows.first.interestDue, 10000);
      expect(rows.first.totalDue, 110000);
    });

    test('reducing interest declines and costs less than flat', () {
      final flat = VikobaCalc.amortize(
          principal: 100000, ratePercent: 10, termMonths: 3);
      final reducing = VikobaCalc.amortize(
          principal: 100000,
          ratePercent: 10,
          termMonths: 3,
          method: LoanInterestMethod.reducing);

      expect(sumPrincipal(reducing), 100000);
      expect(reducing[0].interestDue, greaterThan(reducing[1].interestDue));
      expect(reducing[1].interestDue, greaterThan(reducing[2].interestDue));
      final flatInterest = flat.fold(0.0, (s, r) => s + r.interestDue);
      final reducingInterest =
          reducing.fold(0.0, (s, r) => s + r.interestDue);
      expect(reducingInterest, lessThan(flatInterest));
    });

    test('reducing zero rate is an even principal split', () {
      final rows = VikobaCalc.amortize(
          principal: 120000,
          ratePercent: 0,
          termMonths: 3,
          method: LoanInterestMethod.reducing);
      expect(rows.first.principalDue, 40000);
      expect(rows.first.interestDue, 0);
    });
  });

  group('quote', () {
    test('exposes totals and schedule from a product', () {
      final quote = VikobaCalc.quote(
          product: product(), principal: 100000, termMonths: 3);

      expect(quote.totalInterest, 30000);
      expect(quote.totalPayable, 130000);
      expect(quote.schedule.length, 3);
      expect(quote.method, LoanInterestMethod.flat);
      expect(quote.monthlyInstallment, closeTo(43334, 0.001));
    });
  });

  group('penalties (mirrors backend)', () {
    test('flat 5000 per 7-day period after grace, no grace', () {
      final l = loan(lateBy: const Duration(days: 15), prod: product());
      // ceil(15/7) = 3 charges x 5000.
      expect(
          VikobaCalc.accruedPenalty(loan: l, product: product()),
          closeTo(15000, 0.001));
    });

    test('grace period suppresses the penalty', () {
      final p = product(penaltyGrace: 10);
      final l = loan(lateBy: const Duration(days: 5), prod: p);
      expect(VikobaCalc.accruedPenalty(loan: l, product: p), 0);
    });

    test('no penalty before the due date or for non-active loans', () {
      expect(VikobaCalc.accruedPenalty(
          loan: loan(prod: product()), product: product()),
          0);
      expect(
          VikobaCalc.accruedPenalty(
              loan: loan(
                  status: LoanStatus.pending,
                  lateBy: const Duration(days: 30),
                  prod: product()),
              product: product()),
          0);
    });
  });

  group('repayment waterfall (mirrors backend)', () {
    test('pays penalties then interest then principal in schedule order', () {
      final l = loan(
        prod: product(),
        term: 2,
        lateBy: const Duration(days: 15),
        penaltyAccrued: 15000,
      );
      // balance = 120000 + 15000 = 135000; 50000 payment.
      final split = VikobaCalc.waterfall(loan: l, amount: 50000);

      expect(split.overpayment, 0);
      expect(split.penaltyPaid, 15000);
      expect(split.interestPaid, 10000);
      expect(split.principalPaid, 25000);
    });

    test('overpayment is returned, never credited', () {
      final l = loan(prod: product(), term: 1);
      final split = VikobaCalc.waterfall(loan: l, amount: 150000);

      expect(split.overpayment, 40000);
      expect(split.principalPaid + split.interestPaid, 110000);
      expect(l.balance, 110000);
    });
  });

  group('eligibility (mirrors backend)', () {
    test('applies product rules and membership days', () {
      final p = product();
      // Joined 5 days ago: membership too short.
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(daysJoined: 5),
                  requestedAmount: 100000,
                  allLoans: const [],
                  product: p,
                  termMonths: 3)
              .verdict,
          LoanEligibility.membershipTooShort);

      // Old enough, but 10 shares x 10k x 4 = 400k cap.
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(shares: 10, daysJoined: 120),
                  requestedAmount: 500000,
                  allLoans: const [],
                  product: p,
                  termMonths: 3)
              .verdict,
          LoanEligibility.exceedsMultiple);

      // Term past the product max.
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(shares: 10, daysJoined: 120),
                  requestedAmount: 100000,
                  allLoans: const [],
                  product: p,
                  termMonths: 12)
              .verdict,
          LoanEligibility.termExceedsMax);

      // All good.
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(),
                  requestedAmount: 100000,
                  allLoans: const [],
                  product: p,
                  termMonths: 3)
              .isOk,
          true);
    });

    test('blocks inactive members and repeat borrowers', () {
      final p = product();
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(active: false),
                  requestedAmount: 100000,
                  allLoans: const [],
                  product: p)
              .verdict,
          LoanEligibility.memberInactive);

      final l = loan(prod: p, status: LoanStatus.active);
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(),
                  requestedAmount: 50000,
                  allLoans: [l],
                  product: p)
              .verdict,
          LoanEligibility.alreadyOwing);
    });

    test('below the group minimum is refused', () {
      expect(
          VikobaCalc.checkLoanEligibility(
                  member: member(),
                  requestedAmount: 5000,
                  allLoans: const [],
                  product: product())
              .verdict,
          LoanEligibility.belowMinimum);
    });
  });
}

File: backend/tests/Unit/LoanEngineTest.php
<?php

namespace Tests\Unit;

use App\Models\Group;
use App\Models\Loan;
use App\Models\LoanProduct;
use App\Models\Member;
use App\Models\User;
use App\Services\VikobaService;
use Illuminate\Foundation\Testing\RefreshDatabase;
use Tests\TestCase;

class LoanEngineTest extends TestCase
{
    use RefreshDatabase;

    private Group $group;

    protected function setUp(): void
    {
        parent::setUp();

        $this->group = Group::create([
            'name' => 'Engine Group',
            'share_value' => 10000,
            'default_interest_rate' => 10,
            'max_loan_multiple' => 4,
        ]);
    }

    private function product(string $method = 'flat', float $rate = 10.0): LoanProduct
    {
        return $this->group->loanProducts()->create([
            'name' => 'Test Loan',
            'interest_rate' => $rate,
            'interest_method' => $method,
            'max_term_months' => 6,
            'max_multiplier' => 4,
            'min_amount' => 20000,
            'penalty_type' => 'flat',
            'penalty_value' => 5000,
            'penalty_grace_days' => 0,
            'penalty_period_days' => 7,
        ]);
    }

    private function member(int $shares = 40, int $daysJoined = 120): Member
    {
        return $this->group->members()->create([
            'full_name' => 'Amina Juma',
            'phone' => '0712345678',
            'role' => 'member',
            'joined_date' => now()->subDays($daysJoined)->toDateString(),
            'total_shares' => $shares,
            'share_value' => 10000,
        ]);
    }

    public function test_flat_amortization_sums_exactly(): void
    {
        $rows = VikobaService::amortize(100000, 10.0, 3, 'flat');

        $this->assertCount(3, $rows);
        $this->assertEquals(100000, array_sum(array_column($rows, 'principal')));
        $this->assertEquals(30000, array_sum(array_column($rows, 'interest'))); // 10% x 3 months
        $this->assertEquals(130000, array_sum(array_column($rows, 'total')));
        $this->assertEquals(0, end($rows)['balance_after']);
        // Evenly split with the +1 shilling drift pushed to the earliest rows.
        $this->assertEquals(33334, $rows[0]['principal']);
        $this->assertEquals(33333, $rows[1]['principal']);
        $this->assertEquals(33333, $rows[2]['principal']);
    }

    public function test_single_installment_flat_matches_legacy(): void
    {
        $rows = VikobaService::amortize(100000, 10.0, 1, 'flat');

        $this->assertEquals(10000, $rows[0]['interest']);
        $this->assertEquals(110000, $rows[0]['total']);
    }

    public function test_reducing_amortization_declines_interest_over_principal(): void
    {
        $rows = VikobaService::amortize(100000, 10.0, 3, 'reducing');

        $this->assertCount(3, $rows);
        $this->assertEquals(100000, array_sum(array_column($rows, 'principal')));
        $this->assertEquals(0, end($rows)['balance_after']);
        // Interest declines period over period (10% of a shrinking balance).
        $this->assertGreaterThan($rows[1]['interest'], $rows[0]['interest']);
        $this->assertGreaterThan($rows[2]['interest'], $rows[1]['interest']);
        // Reducing-balance always costs less than flat on the same rate.
        $flatInterest = array_sum(array_column(VikobaService::amortize(100000, 10.0, 3, 'flat'), 'interest'));
        $reducingInterest = array_sum(array_column($rows, 'interest'));
        $this->assertLessThan($flatInterest, $reducingInterest);
    }

    public function test_reducing_zero_rate_is_an_even_principal_split(): void
    {
        $rows = VikobaService::amortize(120000, 0.0, 3, 'reducing');

        $this->assertEquals(40000, $rows[0]['principal']);
        $this->assertEquals(0, $rows[0]['interest']);
        $this->assertEquals(0, end($rows)['balance_after']);
    }

    public function test_quote_matches_amortization_and_exposes_totals(): void
    {
        $quote = VikobaService::quote($this->product('flat', 10), 100000, 3);

        $this->assertEquals(30000, $quote['total_interest']);
        $this->assertEquals(130000, $quote['total_payable']);
        $this->assertCount(3, $quote['schedule']);
        $this->assertEquals('flat', $quote['interest_method']);
    }

    public function test_penalty_accrual_flat_and_percent(): void
    {
        $loan = $this->group->loans()->create([
            'member_id' => $this->member()->id,
            'loan_product_id' => $this->product('flat', 10)->id,
            'principal' => 100000,
            'interest_rate' => 10,
            'interest_method' => 'flat',
            'term_months' => 1,
            'interest_amount' => 10000,
            'total_payable' => 110000,
            'status' => Loan::STATUS_ACTIVE,
            'due_date' => now()->subDays(15)->toDateString(),
        ]);

        // 15 days late, grace 0, period 7 -> ceil(15/7) = 3 charges x 5000.
        $this->assertEquals(15000, VikobaService::accruedPenalty($loan));

        $added = VikobaService::syncPenalties($loan);
        $this->assertEquals(15000, $added);
        $this->assertEquals(15000, $loan->fresh()->penalty_accrued);
        $this->assertCount(3, $loan->fresh()->penalties);
        $this->assertEquals(125000, $loan->fresh()->balance);

        // Running again adds nothing.
        $this->assertEquals(0, VikobaService::syncPenalties($loan));
    }

    public function test_penalty_respects_grace_period(): void
    {
        $loan = $this->group->loans()->create([
            'member_id' => $this->member()->id,
            'loan_product_id' => $this->product()->id,
            'principal' => 100000,
            'interest_rate' => 10,
            'interest_method' => 'flat',
            'term_months' => 1,
            'interest_amount' => 10000,
            'total_payable' => 110000,
            'status' => Loan::STATUS_ACTIVE,
            'due_date' => now()->subDays(5)->toDateString(),
        ]);
        $loan->product->update(['penalty_grace_days' => 10]);

        $this->assertEquals(0, VikobaService::accruedPenalty($loan));
    }

    public function test_repayment_waterfall_pays_penalties_then_interest_then_principal(): void
    {
        $loan = $this->group->loans()->create([
            'member_id' => $this->member()->id,
            'loan_product_id' => $this->product()->id,
            'principal' => 100000,
            'interest_rate' => 10,
            'interest_method' => 'flat',
            'term_months' => 2,
            'interest_amount' => 20000,
            'total_payable' => 120000,
            'status' => Loan::STATUS_ACTIVE,
            'due_date' => now()->subDays(15)->toDateString(),
        ]);
        VikobaService::generateSchedule($loan, $loan->due_date->copy()->subDays(60));
        VikobaService::syncPenalties($loan); // 15000 accrued

        $result = VikobaService::repay($loan, 50000, $this->treasurer()->id);

        $repayment = $result['repayment'];
        $this->assertEquals(0, $result['overpayment']);
        $this->assertEquals(15000, $repayment->penalty_paid);
        // Waterfall walks the schedule in order: first installment's 10k
        // interest, then as much principal as remains (25k into the 50k due).
        $this->assertEquals(10000, $repayment->interest_paid);
        $this->assertEquals(25000, $repayment->principal_paid);

        $fresh = $loan->fresh();
        $this->assertEquals(0, $fresh->penalty_accrued);
        $this->assertEquals(35000, $fresh->amount_repaid);
        $this->assertEquals(85000, $fresh->balance);

        // First schedule has its interest settled and is half-principal paid;
        // the second is untouched.
        $first = $fresh->schedules->first();
        $this->assertEquals(10000, $first->paid_interest);
        $this->assertEquals(25000, $first->paid_principal);
        $this->assertEquals(0, $fresh->schedules->last()->paid_principal);
    }

    public function test_repayment_overpayment_is_returned_not_credited(): void
    {
        $loan = $this->group->loans()->create([
            'member_id' => $this->member()->id,
            'loan_product_id' => $this->product()->id,
            'principal' => 100000,
            'interest_rate' => 10,
            'interest_method' => 'flat',
            'term_months' => 1,
            'interest_amount' => 10000,
            'total_payable' => 110000,
            'status' => Loan::STATUS_ACTIVE,
            'due_date' => now()->addDays(30)->toDateString(),
        ]);
        VikobaService::generateSchedule($loan, now()->subDays(1));

        $result = VikobaService::repay($loan, 150000, $this->treasurer()->id);

        $this->assertEquals(40000, $result['overpayment']);
        $this->assertEquals(110000, $result['repayment']->amount);
        $this->assertEquals(Loan::STATUS_REPAID, $loan->fresh()->status);
    }

    public function test_eligibility_uses_product_rules_and_membership(): void
    {
        $product = $this->product('flat', 10);
        $member = $this->member(shares: 10, daysJoined: 5); // too new, low savings

        // Membership too short.
        $check = VikobaService::loanEligibility($member, $this->group, 100000, [], $product, 3);
        $this->assertFalse($check['ok']);
        $this->assertEquals('membership_too_short', $check['reason']);

        // Old enough now, but over the product multiplier (10 shares x 10k x 4 = 400k).
        $member->update(['joined_date' => now()->subDays(120)->toDateString()]);
        $check = VikobaService::loanEligibility($member, $this->group, 500000, [], $product, 3);
        $this->assertFalse($check['ok']);
        $this->assertEquals('exceeds_savings_multiple', $check['reason']);

        // Term exceeds product max.
        $check = VikobaService::loanEligibility($member, $this->group, 100000, [], $product, 12);
        $this->assertFalse($check['ok']);
        $this->assertEquals('term_exceeds_max', $check['reason']);

        // All good.
        $check = VikobaService::loanEligibility($member, $this->group, 100000, [], $product, 3);
        $this->assertTrue($check['ok']);
    }

    public function test_eligibility_blocks_when_already_owing(): void
    {
        $member = $this->member();
        $this->group->loans()->create([
            'member_id' => $member->id,
            'principal' => 50000,
            'interest_rate' => 10,
            'interest_amount' => 5000,
            'total_payable' => 55000,
            'status' => Loan::STATUS_ACTIVE,
            'due_date' => now()->addDays(30)->toDateString(),
        ]);

        $check = VikobaService::loanEligibility($member, $this->group, 50000, $this->group->loans()->get(), $this->product(), 3);
        $this->assertFalse($check['ok']);
        $this->assertEquals('already_owing', $check['reason']);
    }

    private function treasurer(): User
    {
        return User::create([
            'name' => 'Mweka Hazina',
            'email' => 'treasurer@engine.test',
            'phone' => '0765432109',
            'password' => bcrypt('password'),
            'role' => User::ROLE_TREASURER,
            'group_id' => $this->group->id,
        ]);
    }
}
